### ISSUE TITLE
midi-qol doesn't wait on "OnUseMacro" to finish before auto-rolling attack and damage rolls

### ISSUE THREAD

**Issue by tposney** [tposney/midi-qol#934]
In GitLab by @wickermoon on Aug 4, 2022, 09:46

I have an item, a class feature, which is setup this way:

![image](https://my-gitlab-bucket.s3.amazonaws.com/640711674/48c958b63bc3a7a07b3f3c34615cfd940581c135347b55a1b2d7c07793d3ec10/image.png)

The On Use Macros-script is pretty straight-forward:

![image](https://my-gitlab-bucket.s3.amazonaws.com/640711674/9ab4f86464f9cea871ef0636901ef06f3be8f7df583d09aa8caf7519a07ffa80/image.png)

It simply opens a dialog which prompts the user to specify the number of dice to use.

This is necessary, as the feature in question has variable healing-damage, depending on the dice the user spends.

I have also set the following options in the midi-qol settings:

![image](https://my-gitlab-bucket.s3.amazonaws.com/640711674/291c1194994dbeba3760981bef49e0eafff29c5c5a0a0819452bc7e4809e4563/image.png)

I don't want my players to always have to click stuff.

Now, I was of the impression that the auto-fast-forward rolls for damage (and attack) would wait for any on-use-macros script to finish. But apparently it does not.
When I activate the ability, I get the usage configuration:

![image](https://my-gitlab-bucket.s3.amazonaws.com/640711674/d7bfde5a645da4959083f3895e22155f9b6cccd2c692814808b62a6b434e9817/image.png)

and the script dialog:

 ![image](https://my-gitlab-bucket.s3.amazonaws.com/640711674/540d9a37738cb5b724fb4cc7d47332fb6bf61d73674b3a69b4cdcb92ac051c22/image.png)

(the latter being obscured by the usage configuration dialog)

And as soon as I close the usage configuration dialog, both attack and damage rolls are being made. I think that is a bug. But if it is not, I would say that there's a feature missing: A blocking script.

**Comment by tposney** [comment]
Can you attach the item to the report and I'll have a look. The intention is to wait for the onUseMacro to complete, I wonder if you macro is not returning a promise for me to await.
Possibly a return dialog/await dialog will fix it?

**Comment by tposney** [comment]
In GitLab by @wickermoon on Aug 5, 2022, 20:03

Here's the script in copy-able form and the corresponding item which uses it. I tried both, returning the dialog and awaiting the dialog, but that won't fix the problem. :/

```js
const content =
` <form>
    <div class="form-group">
      <label> Number of dice: </label>
      <input type="number" id="diceNumber" placeholder="e.g. 2" />
    </div>
  </form>
`

const dialog = new Dialog({
  title: "Healing Light: Usage Configuration II",
  content: content,
  buttons:
  {
    button1:
    {
      label: "Heal!",
      callback: (html) => OnHeal(html)
    },
    button2:
    {
      label: "Cancel",
      callback: () => { }
    }
  },
  default: "button1"
}).render(true);


//ui.notifications.info(`Number of Dice: ${dice}`);
//const clone = healingLight.clone({"data.uses.per":""}, {keepId: true});
//await clone.roll({configureDialog:false});
async function OnHeal(html)
{
  const input = Math.max(1, html.find("input#diceNumber").val());
  const max = token.actor.data.data.abilities.cha.mod;

  let dice = Math.min(input, max);
  let healingLight = token.actor.items.getName("Healing Light");
  let remaining_uses = healingLight.toObject().data.uses.value - dice;

  if(remaining_uses < 0)
  {
    dice += remaining_uses;
    remaining_uses = 0;
  }

  let parts = [[`${dice}@scale.the-celestial.healing-light-die`, "healing"]];
  await healingLight.update({"data.damage.parts":parts});
  await healingLight.update({"data.uses.value":remaining_uses });
}
```

[fvtt-Item-healing-light.json](https://my-gitlab-bucket.s3.amazonaws.com/640711674/bf8582748747be78d074ac73466165e12a4ffce38839f39eb8da733a2b1a37da/fvtt-Item-healing-light.json)

**Comment by tposney** [comment]
In GitLab by @wickermoon on Aug 5, 2022, 20:45

Okay, I've wrapped the dialog in a return new Promise((resolve, reject) => { ... }); statement and now the whole shabang waits for my dialog to finish, which is perfect. I thought it always returned a promise, but then again, this is my first script. Sorry for the bug report! Now I have to resolve the problem of why it doesn't execute the damage roll after.

**Comment by tposney** [comment]
There's one wrinkle in the macro that you need to worry about. Since OneHeal is async it may complete AFTER the item roll completes (there'd have to be something else to interrupt processing, but it could happen).

If you code it like this
```js
const content =
` <form>
    <div class="form-group">
      <label> Number of dice: </label>
      <input type="number" id="diceNumber" placeholder="e.g. 2" />
    </div>
  </form>
`

return new Promise(resolve => {
  async function OnHeal(html)
  {
    const input = Math.max(1, html.find("input#diceNumber").val());
    const max = token.actor.data.data.abilities.cha.mod;

    let dice = Math.min(input, max);
    let healingLight = await fromUuid(args[0].itemUuid);
    let remaining_uses = healingLight.data.data.uses.value - dice;

    if(remaining_uses < 0)
    {
      dice += remaining_uses;
      remaining_uses = 0;
    }

    let parts = [[`${dice}@scale.the-celestial.healing-light-die`, "healing"]];
    await healingLight.update({"data.damage.parts":parts, "data.uses.value": remaining_uses}); 
    // only 1 database transaction so better
    resolve(true);
  }
  new Dialog({
    title: "Healing Light: Usage Configuration II",
    content: content,
    buttons:
    {
      button1:
      {
        label: "Heal!",
        callback: OnHeal
      },
      button2:
      {
        label: "Cancel",
        callback: () => { resolve(true) }
      }
    },
    default: "button1"
  }).render(true);
});
```

Then the promise won't resolve until the updates have finished.

**Comment by tposney** [comment]
In GitLab by @wickermoon on Aug 5, 2022, 22:39

Super cool, thank you so much for the help! I'll close the issue, since it's not really an issue, but just me not understanding how the stuff works. But one last question: I don't want to use "resolve(true);" since it pops up the Usage Configuration Dialog. I'd rather just show the feature-card in the side-bar and trigger rolling damage. I know that I can just call "healingLight.rollDamage()" to roll for the damage, but neither will the feature info-box be shown, nor will the players see the outcome of the roll. Is there a way to manually trigger that? I know that's not part of the issue, so if you don't want to bother with that, that's absolutely understandable!

Again, thanks a bunch!

**Comment by tposney** [comment]
There is, but it's a little convoluted. Basically you set the workflow to auto consume resources (which will use up one charge) then tweak the resources remaining to account for the fact that a charge will get used.

It also has the bonus that the you don't have any uses left notification will trigger if appropriate.

I also tweak a few little places in the code
```js
console.error("args are", args);
const content =
` <form>
    <div class="form-group">
      <label> Number of dice: </label>
      <input type="number" id="diceNumber" placeholder="e.g. 2" />
    </div>
  </form>
`

return new Promise(resolve => {
  async function OnHeal(html)
  {
    const input = Math.max(1, html.find("input#diceNumber").val());
    const max = token.actor.data.data.abilities.cha.mod;

    let dice = Math.min(input, max);
    let healingLight = await fromUuid(args[0].itemUuid);
    let remaining_uses = healingLight.data.data.uses.value - dice + 1;

    if(remaining_uses < 1)
    {
      dice += remaining_uses - 1;
      remaining_uses = dice > 0 ? 1 : 0;
    }

    let parts = [[`${dice}@scale.the-celestial.healing-light-die`, "healing"]];
    // only 1 database transaction so better
    await healingLight.update({"data.damage.parts":parts, "data.uses.value": remaining_uses}); 

    // update the workflow to autoConsume the resource (no dialog)
    // we have to fetch the live workflow and change it for this to work
    let workflow = MidiQOL.Workflow.getWorkflow(args[0].uuid);
    workflow.workflowOptions.autoConsumeResource = "both";
    resolve(true);
  }
  new Dialog({
    title: "Healing Light: Usage Configuration II",
    content: content,
    buttons:
    {
      button1:
      {
        label: "Heal!",
        callback: OnHeal
      },
      button2:
      {
        label: "Cancel",
        callback: () => { resolve(true) }
      }
    },
    default: "button1"
  }).render(true);
});
```

**Comment by tposney** [comment]
In GitLab by @wickermoon on Aug 5, 2022, 23:17

omfg, this is so awesome! Thanks a ton! :D So happy right now!
